docs/content/guides/dagster/pythonic-resources.mdx

```diff
@@ -0,0 +1,337 @@
+
+
+class CredentialsResource(ConfigurableResource):
+    username: str
+    password: str
+
+
+class FileStoreBucket(ConfigurableResource):
+    credentials: CredentialsResource
+    region: str
+
+    def write(self, data: str):
+        get_filestore_client(
+            username=self.credentials.username,
+            password=self.credentials.password,
+            region=self.region,
+        ).write(data)
+
+
+credentials = CredentialsResource(username="my_user", password="my_password")
+defs = Definitions(
+    assets=...,
+    resources={
+        "bucket": FileStoreBucket(
+            credentials=credentials,
+            region="us-east-1",
+        ),
+    },
+)
+```
+
+If we instead would like the configuration for our credentials to be provided at runtime, we can use the `configure_at_launch()` method to defer the construction of the `CredentialsResource` until runtime.
+
+Because `credentials` requires runtime configuration through the launchpad, it must also be passed to the `Definitions` object, so that configuration can be provided at runtime. Nested resources only need to be passed to the `Definitions` object if they require runtime configuration.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_dep_job_runtime endbefore=end_new_resource_dep_job_runtime
+credentials = CredentialsResource.configure_at_launch()
+
+
+defs = Definitions(
+    assets=...,
+    resources={
+        "credentials": credentials,
+        "bucket": FileStoreBucket(
+            credentials=credentials,
+            region="us-east-1",
+        ),
+    },
+)
+```
+
+---
+
+## Testing configurable resources
+
+You can test the initialization of a <PyObject object="ConfigurableResource"/> by constructing it manually. In most cases, you can construct your resource directly:
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_testing endbefore=end_new_resource_testing
+from dagster._config.structured_config import ConfigurableResource
+
+
+class MyResource(ConfigurableResource):
+    def get_value(self) -> str:
+        return "foo"
+
+
+def test_my_resource():
+    assert MyResource().get_value() == "foo"
+```
+
+If your resource requires other resources, then you can pass them as constructor arguments.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_new_resource_testing_with_context endbefore=end_new_resource_testing_with_context
+from dagster._config.structured_config import ConfigurableResource
+
+
+class StringHolderResource(ConfigurableResource):
+    value: str
+
+
+class MyResourceRequiresAnother(ConfigurableResource):
+    foo: StringHolderResource
+    bar: str
+
+
+def test_my_resource_with_context():
+    resource = MyResourceRequiresAnother(foo=StringHolderResource("foo"), bar="bar")
+    assert resource.foo.value == "foo"
+    assert resource.bar == "bar"
+```
```

Comment:
Let's add a section describe typing and how we accept different types at the constructor than the output. E.g. you can pass `EnvVar` in but you always get `str` out.

Reply:
Added in an `implementation details` section